File: backend/services/collector.py
"""
Naver blog RSS collection and HTML scraping.
Supports ordered_blocks (text + image interleaved) for LLM extraction.
"""
from __future__ import annotations
import re, time
from typing import Callable, List, Optional, Tuple
import requests
from bs4 import BeautifulSoup

from models.schemas import OrderedBlock, PostItem, ScrapedPostData

# ── Blog list (blogId, categoryFolder) ───────────────────────────────────────
BLOGS: List[Tuple[str, str]] = [
    ("hsh6566",     "쇼핑"),
    ("hkh443",      "코스트코"),
    ("hprbel1097",  "방송아이템"),
    ("jsodnfak",    "먹거리"),
    ("fashionblog", "파트너스활동으로 소정 수익발생"),
    ("bcf5qp11",    "쇼핑"),
    ("celubdigging", "파트너스활동으로 소정 수익발생"),
    ("dalcome5",    "◇궁금해◇"),
    ("greenp77",    "쇼핑"),
    ("jiyeon_style","패션"),
    ("beauty_daily","뷰티"),
    ("kdrama_item", "방송아이템"),
    ("celeb_pick",  "셀럽픽"),
    ("stylenote_kr","스타일"),
    ("kfashion_lab","패션"),
]

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120 Safari/537.36"
}


# ── RSS collection ────────────────────────────────────────────────────────────

def _get_rss(name: str, folder: str) -> List[PostItem]:
    try:
        import xmltodict
        resp = requests.get(
            f"https://rss.blog.naver.com/{name}.xml",
            headers=HEADERS, timeout=8, verify=False,
        )
        resp.raise_for_status()
        doc = xmltodict.parse(resp.text)
        items = doc.get("rss", {}).get("channel", {}).get("item", [])
        if isinstance(items, dict):
            items = [items]
        result: List[PostItem] = []
        for it in items:
            cats = it.get("category", "")
            if isinstance(cats, list):
                cats = " ".join(cats)
            if folder.lower() not in str(cats).lower():
                continue
            guid = it.get("guid", "")
            url = (guid.get("#text", "") if isinstance(guid, dict) else guid) or ""
            result.append(PostItem(
                title=str(it.get("title", "")),
                url=str(url),
                date=str(it.get("pubDate", "")),
                tag=str(it.get("category", "")),
            ))
        return result
    except Exception:
        return []


def collect_posts(
    days: int = 2,
    on_progress: Optional[Callable[[int, int], None]] = None,
) -> List[PostItem]:
    import urllib3
    urllib3.disable_warnings()
    cutoff = time.time() - days * 86400
    all_posts: List[PostItem] = []
    total = len(BLOGS)

    for i, (name, folder) in enumerate(BLOGS):
        posts = _get_rss(name, folder)
        all_posts.extend(posts)
        if on_progress:
            on_progress(i + 1, total)

    valid: List[PostItem] = []
    for p in all_posts:
        try:
            from email.utils import parsedate_to_datetime
            t = parsedate_to_datetime(p.date).timestamp()
            if t >= cutoff:
                valid.append(p)
        except Exception:
            pass
    valid.sort(key=lambda p: p.date, reverse=True)
    return valid


# ── HTML scraping ─────────────────────────────────────────────────────────────

def _parse_naver_url(url: str) -> Optional[Tuple[str, str]]:
    m = re.search(r'blog\.naver\.com/([^/?#]+)/(\d+)', url)
    if not m:
        m = re.search(r'blogId=([^&]+).*logNo=(\d+)', url)
    return (m.group(1), m.group(2)) if m else None


def _build_postview_url(blog_id: str, log_no: str) -> str:
    return (
        f"https://blog.naver.com/PostView.naver"
        f"?blogId={blog_id}&logNo={log_no}"
        f"&redirect=Dlog&widgetTypeCall=true&directAccess=false"
    )


def _fetch_html(url: str, timeout: int = 12) -> str:
    import urllib3
    urllib3.disable_warnings()
    resp = requests.get(url, headers=HEADERS, timeout=timeout, verify=False)
    resp.raise_for_status()
    return resp.text


def _parse_naver_html(html: str, title: str, post_url: str) -> ScrapedPostData:
    soup = BeautifulSoup(html, "lxml")

    # Images
    image_urls: List[str] = []
    seen_imgs: set[str] = set()
    for img in soup.select("img.se-image-resource, img[src*='postfiles.pstatic.net']"):
        src = img.get("src") or img.get("data-lazy-src") or ""
        if src and "pstatic.net" in src:
            src = re.sub(r'\?type=\w+', '?type=w966', src)
            if src not in seen_imgs:
                seen_imgs.add(src)
                image_urls.append(src)

    # Paragraphs
    paragraphs: List[str] = []
    for p in soup.select("p.se-text-paragraph"):
        t = (p.get_text() or "").replace("\u200b", "").replace("\xa0", " ").strip()
        if t:
            paragraphs.append(t)

    # Ordered blocks
    ordered_blocks: List[OrderedBlock] = []
    container = soup.select_one("div.se-main-container, div.__se_component_area")
    if container:
        def walk(node):
            cls = node.get("class") or []
            if node.name == "p" and "se-text-paragraph" in cls:
                t = (node.get_text() or "").replace("\u200b", "").replace("\xa0", " ").strip()
                if t:
                    ordered_blocks.append(OrderedBlock(type="text", content=t))
            elif node.name == "img" and (
                "se-image-resource" in cls
                or "postfiles.pstatic.net" in (node.get("src") or "")
            ):
                src = node.get("src") or node.get("data-lazy-src") or ""
                if src and "pstatic.net" in src:
                    src = re.sub(r'\?type=\w+', '?type=w966', src)
                    ordered_blocks.append(OrderedBlock(type="image", url=src))
            else:
                for child in node.children:
                    if hasattr(child, "name") and child.name:
                        walk(child)

        for child in container.children:
            if hasattr(child, "name") and child.name:
                walk(child)

    # Links — include all known short-URL domains + coupang / smartstore
    _LINK_KEYWORDS = (
        "coupang", "smartstore", "naver.me",
        "vvd.bz", "han.gl", "me2.do", "url.kr",
        "bit.ly", "t.co", "tinyurl.com", "ow.ly", "goo.gl",
        "smarturl.it", "rebrand.ly", "buff.ly", "cutt.ly",
        "vo.la", "c11.kr", "mrk.kr", "lrl.kr", "glink.page",
        "coupa.ng", "link.coupang.com",
    )
    _selector = ", ".join(
        f"a[href*='{kw}']" for kw in _LINK_KEYWORDS
    ) + ", a.se-link"
    links: List[dict] = []
    seen_hrefs: set[str] = set()
    for a in soup.select(_selector):
        href = a.get("href") or ""
        text = (a.get_text() or "").strip()
        if href.startswith("http") and href not in seen_hrefs:
            seen_hrefs.add(href)
            links.append({"text": text, "href": href})

    return ScrapedPostData(
        ordered_blocks=ordered_blocks,
        image_urls=image_urls,
        paragraphs=paragraphs,
        links=links,
        post_url=post_url,
        title=title,
    )


def scrape_post(post: PostItem) -> Optional[ScrapedPostData]:
    try:
        parsed = _parse_naver_url(post.url)
        if not parsed:
            return None
        pv_url = _build_postview_url(*parsed)
        html = _fetch_html(pv_url)
        data = _parse_naver_html(html, post.title, post.url)
        return data if (data.paragraphs or data.image_urls) else None
    except Exception:
        return None


def scrape_multiple_posts(
    posts: List[PostItem],
    max_posts: int = 10,
    on_progress: Optional[Callable[[int, int], None]] = None,
) -> List[ScrapedPostData]:
    targets = posts[:max_posts]
    results: List[ScrapedPostData] = []
    for i, post in enumerate(targets):
        data = scrape_post(post)
        if data:
            results.append(data)
        if on_progress:
            on_progress(i + 1, len(targets))
    return results

File: standalone/src/lib/collector.ts
import { BLOGS } from "./blogs";
import type { OrderedBlock, PostItem, ScrapedPostData } from "./types";

// ── CORS 프록시 (순서대로 fallback) ──────────────────────────────
const PROXIES = [
  (url: string) => `https://api.allorigins.win/raw?url=${encodeURIComponent(url)}&_=${Date.now()}`,
  (url: string) => `https://corsproxy.io/?${encodeURIComponent(url)}`,
];

async function fetchViaProxy(url: string, timeoutMs = 10000): Promise<string> {
  let lastErr: unknown;
  for (const proxyFn of PROXIES) {
    try {
      const ctrl = new AbortController();
      const tid = setTimeout(() => ctrl.abort(), timeoutMs);
      const res = await fetch(proxyFn(url), { signal: ctrl.signal });
      clearTimeout(tid);
      if (res.ok) {
        const text = await res.text();
        if (text.length > 100) return text;
      }
    } catch (e) {
      lastErr = e;
    }
  }
  throw lastErr ?? new Error("All proxies failed");
}

// ── RSS 수집 ──────────────────────────────────────────────────────
function parseRSS(xml: string, folder: string): PostItem[] {
  try {
    const doc = new DOMParser().parseFromString(xml, "application/xml");
    return Array.from(doc.querySelectorAll("item"))
      .filter((el) => el.querySelector("category")?.textContent?.trim() === folder)
      .map((el) => ({
        title: el.querySelector("title")?.textContent ?? "",
        url:   el.querySelector("guid")?.textContent ?? "",
        date:  el.querySelector("pubDate")?.textContent ?? "",
        tag:   el.querySelector("tag")?.textContent ?? "",
      }));
  } catch {
    return [];
  }
}

async function fetchRSS(name: string, folder: string): Promise<PostItem[]> {
  try {
    const xml = await fetchViaProxy(`https://rss.blog.naver.com/${name}.xml`);
    return parseRSS(xml, folder);
  } catch {
    return [];
  }
}

export async function collectPosts(
  days: number,
  onProgress?: (done: number, total: number) => void
): Promise<PostItem[]> {
  const results: PostItem[] = [];
  const CHUNK = 3;

  for (let i = 0; i < BLOGS.length; i += CHUNK) {
    const chunk = BLOGS.slice(i, i + CHUNK);
    const chunkRes = await Promise.all(
      chunk.map(async ([name, folder]) => {
        const posts = await fetchRSS(name, folder);
        onProgress?.(Math.min(i + CHUNK, BLOGS.length), BLOGS.length);
        return posts;
      })
    );
    results.push(...chunkRes.flat());
  }

  const cutoff = Date.now() - days * 24 * 60 * 60 * 1000;
  return results
    .filter((p) => {
      const t = new Date(p.date).getTime();
      return !isNaN(t) && t >= cutoff;
    })
    .sort((a, b) => new Date(b.date).getTime() - new Date(a.date).getTime());
}

// ── 블로그 HTML 스크래핑 ──────────────────────────────────────────

/** 네이버 블로그 URL → blogId + logNo 추출 */
function parseNaverUrl(url: string): { blogId: string; logNo: string } | null {
  const m =
    url.match(/blog\.naver\.com\/([^/?#]+)\/(\d+)/) ??
    url.match(/blogId=([^&]+).*logNo=(\d+)/);
  if (!m) return null;
  return { blogId: m[1], logNo: m[2] };
}

function buildPostViewUrl(blogId: string, logNo: string): string {
  return `https://blog.naver.com/PostView.naver?blogId=${blogId}&logNo=${logNo}&redirect=Dlog&widgetTypeCall=true&directAccess=false`;
}

function parseNaverHtml(html: string, sourceTitle: string, sourceUrl: string): ScrapedPostData {
  const doc = new DOMParser().parseFromString(html, "text/html");

  // ── 이미지 URL (원본 해상도) ──
  const imageUrls: string[] = [];
  const seenImgs = new Set<string>();
  doc.querySelectorAll("img.se-image-resource, img[src*='postfiles.pstatic.net']").forEach((img) => {
    let src = img.getAttribute("src") || img.getAttribute("data-lazy-src") || "";
    if (src && src.includes("pstatic.net")) {
      src = src.replace(/\?type=\w+/, "?type=w966");
      if (!seenImgs.has(src)) { seenImgs.add(src); imageUrls.push(src); }
    }
  });

  // ── 텍스트 단락 ──
  const paragraphs: string[] = [];
  doc.querySelectorAll("p.se-text-paragraph").forEach((p) => {
    const t = p.textContent?.replace(/\u200b/g, "").replace(/\xa0/g, " ").trim() ?? "";
    if (t) paragraphs.push(t);
  });

  // ── ordered_blocks (텍스트+이미지 순서 보존) ──
  const orderedBlocks: OrderedBlock[] = [];
  const container = doc.querySelector("div.se-main-container, div.__se_component_area");
  if (container) {
    const walk = (node: Element) => {
      if (node.matches("p.se-text-paragraph")) {
        const t = node.textContent?.replace(/\u200b/g, "").replace(/\xa0/g, " ").trim() ?? "";
        if (t) orderedBlocks.push({ type: "text", content: t });
      } else if (node.matches("img.se-image-resource, img[src*='postfiles.pstatic.net']")) {
        let src = node.getAttribute("src") || node.getAttribute("data-lazy-src") || "";
        if (src && src.includes("pstatic.net")) {
          src = src.replace(/\?type=\w+/, "?type=w966");
          orderedBlocks.push({ type: "image", url: src });
        }
      } else {
        Array.from(node.children).forEach(walk);
      }
    };
    Array.from(container.children).forEach(walk);
  }

  // ── 링크 ──
  const links: Array<{ text: string; href: string }> = [];
  doc.querySelectorAll("a.se-link, a[href*='coupang'], a[href*='smartstore'], a[href*='vvd.bz']").forEach((a) => {
    const href = a.getAttribute("href") || "";
    const text = a.textContent?.trim() ?? "";
    if (href.startsWith("http")) links.push({ text, href });
  });

  return { orderedBlocks, imageUrls, paragraphs, links, postUrl: sourceUrl, title: sourceTitle };
}

export async function scrapePost(post: PostItem): Promise<ScrapedPostData | null> {
  try {
    const parsed = parseNaverUrl(post.url);
    if (!parsed) return null;
    const pvUrl = buildPostViewUrl(parsed.blogId, parsed.logNo);
    const html = await fetchViaProxy(pvUrl, 12000);
    return parseNaverHtml(html, post.title, post.url);
  } catch {
    return null;
  }
}

export async function scrapeMultiplePosts(
  posts: PostItem[],
  maxPosts = 8,
  onProgress?: (done: number, total: number) => void
): Promise<ScrapedPostData[]> {
  const targets = posts.slice(0, maxPosts);
  const results: ScrapedPostData[] = [];

  for (let i = 0; i < targets.length; i++) {
    const data = await scrapePost(targets[i]);
    if (data && (data.paragraphs.length > 0 || data.imageUrls.length > 0)) {
      results.push(data);
    }
    onProgress?.(i + 1, targets.length);
  }

  return results;
}
